util/pkg/vfs/vaultcontext.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vfs
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	vault "github.com/hashicorp/vault/api"
+)
+
+func newVaultClient(addr string) (*vault.Client, error) {
+
+	config := &vault.Config{
+		Address:    addr,
+		HttpClient: &http.Client{},
+	}
+	client, err := vault.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, err
+}
+
+func authenticateVaultClient(client *vault.Client) error {
+	token := os.Getenv("VAULT_TOKEN")
+	if token != "" {
+		client.SetToken(token)
+	} else {
+		return awsAuth(client)
```

Comment:
i dont have much experience with vault or other cloud providers, but is this something we could easily extend to support other cloud providers?

Reply:
Sure. We have https://www.vaultproject.io/docs/auth/gcp.html too. It should be quite easy to extend to. Is there an easy way to know the cloud provider in that part of the code? If so, I can add some conditional here. Someone with access to GCP would have to do the actual implementation though.

Reply:
Currently `VFSContext` is implicitly initialized. It expects to be able to get credentials out of the running environment and figures out which cloud provider to use by the protocol part of the input URL. So other cloud providers would either have to use a protocol other than "vault:" or would have to key off of something environmental.

I wonder if `VFSContext` should be explicitly initialized with the `ClusterSpec`. I think the first refactor (for a different PR) would be to remove the global `vfs.Context`.

Reply:
The way I left this now, is to not care about the error from awsauth. So one could attempt auth with gcp or similar if vault still hasn't set a token. It is not ideal, but it will work.

Refactor later on to get cloud provider passed in makes sense, but we have to sync up with etcd-manager here since they also use VFS from kops (there is even logic only used by etcd-manager in kops VFS)